File: exam/03/lab/online_library_system.py
# alternative provided by TA 鄧晉杰
from collections import deque
def Library (m, queries):
    responses = []
    pending_users = {}
    borrowing_books = {}
    for query in queries:
        opcode = query[0]
        if opcode == 1:
            uid = query[1]
            bid = query[2]
            if bid not in pending_users:
                pending_users[bid] = deque()
            pending_users[bid].append(uid)
            if uid not in borrowing_books:
                borrowing_books[uid] = 0
            borrowing_books[uid] += 1
        elif opcode == 2:
            bid = query[1]
            uid = pending_users[bid][0]
            pending_users[bid].popleft()
            borrowing_books[uid] -= 1
        elif opcode == 3:
            bid = query[1]
            if bid in pending_users and len(pending_users) != 0:
                responses.append(pending_users[bid][0])
            else:
                responses.append(0)
        else:
            uid = query[1]
            if uid in borrowing_books:
                responses.append(borrowing_books[uid])
            else:
                responses.append(0)
    return responses

File: exam/03/lab/online_library_system.cc
// alternative provided by TA 鄧晉杰
std::vector<long> Library (long m, std::vector<std::vector<long>> queries)
{
  std::vector<long> responses;
  std::map<long, std::deque<long>> pending_users;
  std::map<long, long> borrowing_books;
  for (auto const &query : queries)
  {
    auto opcode {query[0]};
    if (opcode == 1) // borrow: 1 uid bid
    {
      auto uid {query[1]};
      auto bid {query[2]};
      if (pending_users.find(bid) == pending_users.end())
      {
        pending_users[bid] = std::deque<long>();
      }
      pending_users[bid].push_back(uid);
      if (borrowing_books.find(uid) == borrowing_books.end())
      {
        borrowing_books[uid] = 0;
      }
      ++borrowing_books[uid];
    }
    else if (opcode == 2) // return: 2 bid -1
    {
      auto bid {query[1]};
      auto uid {pending_users[bid].front()};
      pending_users[bid].pop_front();
      --borrowing_books[uid];
    }
    else if (opcode == 3) // book state: 3 bid -1
    {
      auto bid {query[1]};
      if ((pending_users.find(bid) != pending_users.end()) && !pending_users[bid].empty())
      {
        responses.push_back(pending_users[bid].front());
      }
      else
      {
        responses.push_back(0);
      }
    }
    else // user state: 4 uid -1
    {
      auto uid {query[1]};
      if (borrowing_books.find(uid) != borrowing_books.end())
      {
        responses.push_back(borrowing_books[uid]);
      }
      else
      {
        responses.push_back(0);
      }
    }
  }
  return responses;
}
